### ISSUE TITLE
Add random identicons for new users 

### ISSUE THREAD

**Issue by Zigolox** [digitalungdom-se/digitalungdom-api#13]
We should ask if we can use the code created by @adelhult to generate identicons. [The code can be found here](https://eli.nu/BotIdenticon/0.1/mod.js). It was written for deno so we might be able to get the Typescript version.

**Comment by Zigolox** [comment]
Transferred this issue to digitalungdom-api so will close this issue

**Comment by Zigolox** [comment]
@kelszo is it possible to implement this?

**Comment by adelhult** [comment]
I originally thought it would be easy to port my code to ordinary client-side JS, I had hopes that it would work without changing anything, really. But it seems that the bundled file (the file you linked) produced by Deno uses top level await (outside of functions) which I sadly don’t think is supported by modern browsers yet.

However, I have no problem rewriting the code to run on the server side in a Node runtime or client side as React component! The only difference would be plugging my code into another hash-library, it should be a quick fix!

**Comment by adelhult** [comment]
The current code looks something like this:
```js
import { createHash } from "https://deno.land/std/hash/mod.ts";
import faces from './faces.js';

export default class BotIdenticon {
    constructor(identifier) {
        this.identifier = identifier;
        this._hash = this._generateHash(identifier);
        // Hue: values between 0 and 360
        this._hue = (this._hash[0] + this._hash[1]) % 361;
        // Saturation: values between 50% and 100%
        this._saturation = (this._hash[2] % 51) + 50;
        // Lightness: values between 30% and 70%
        this._lightness = (this._hash[3] % 41) + 30;
        this.color = [this._hue, this._saturation, this._lightness];  
        this._face = this._hash[4] % 8;
    }

    _generateHash(identifier) {
        const hash = createHash("md5");
        hash.update(identifier);
        return hash.digest(); 
    }

    get svg() {
        return faces(this._face, this._hue, this._saturation, this._lightness)
            .replace(/\n/gm," ")
            .replace(/\s+/g,' ')
            .trim();
    }
}
``` 
where faces basically is a function that returns a templated js string containing a bunch of svg code.

**Comment by kelszo** [comment]
It is possible to implement this back-end but i recommend implementing it front-end.

**Comment by Nautman** [comment]
I've looked into the SVG:s that Eli have created, and I must say I am impressed beyond what I could have imagined! They are extremely well made. These should definitely be implemented, and I believe that it shouldn't be too difficult!

**Comment by Nautman** [comment]
I believe that a React component is most suitable. If Eli would like to, I could create a base template for a component that would fit in with our current code.

**Comment by adelhult** [comment]
Sure, that would be great! If you design a simple API I could prob. implement it during this week-end. I would imagine something along these lines, but you are free to decide what you would like:
```jsx
<BotIdenticon identifier="Eli Adelhult" />
```

**Comment by Nautman** [comment]
A little update is that we tried using your SVG:s to create a flat logo, that works in both light and dark mode!

<img width="1433" alt="Screenshot 2020-11-19 at 14 27 48" src="https://user-images.githubusercontent.com/28629647/99672230-06ddd280-2a6b-11eb-8ddb-98acf8ff7242.png">


<img width="1437" alt="Screenshot 2020-11-19 at 14 25 52" src="https://user-images.githubusercontent.com/28629647/99672020-c1b9a080-2a6a-11eb-9f74-a5da20849e40.png">

I hope that you think it's okay that we use them :)

**Comment by adelhult** [comment]
Nice! Looks great, feel free to use them for whatever you want :)

**Comment by Nautman** [comment]
If it makes you laugh, this is what my placeholder design looked like
<img width="227" alt="Screenshot 2020-11-19 at 14 43 28" src="https://user-images.githubusercontent.com/28629647/99673909-38579d80-2a6d-11eb-82e0-e1be838a36ee.png">

<img width="223" alt="Screenshot 2020-11-19 at 14 43 47" src="https://user-images.githubusercontent.com/28629647/99673931-41486f00-2a6d-11eb-8577-b074fcefc399.png">

**Comment by Nautman** [comment]
I agree that the identicon should be created on the front-end. However, maybe once we have implemented random parameters for the identicon we could save them in the database. Then people could create a personalized bot!

**Comment by adelhult** [comment]
I had some time over to night so a quickly created a first working draft of the component, I hope you like it! :)

https://github.com/adelhult/BotIdenticon

**Comment by Nautman** [comment]
I like it very much! @adelhult 🥳 . There are some small changes that I would like to make:

1. Convert it to TypeScript, as to conform with the rest of our code and standards.
2. Make a deterministic way of creating a specific identicon, by creating props for `face` and HSL-`color`. The hash function can then be used to determine the props.

**Comment by adelhult** [comment]
@Nautman, those are great suggestions. I just pushed (2). However, I have actually never written any Typescript so I'm afraid I that can't  help with that. But, I can give you all write access to the repo if you would like to convert the code.

**Comment by Nautman** [comment]
Great! I can create a pull request for TypeScript. There is no need for sharing access to the repo, at least yet.

TypeScript isn't so hard, it's basically only about defining what values a variable can take (eg. number, string, a combination of those, map, etc.) and what values functions return.
